File: src/sugarglider/api/saved_routes.py
"""Immutable saved-route persistence and unlisted-sharing endpoints."""

from typing import Annotated

from fastapi import APIRouter, Body, Header, Response, status

from sugarglider.api.dependencies import SavedRouteServiceDependency
from sugarglider.gpx.writer import gpx_filename, write_plan_gpx
from sugarglider.saved_routes.models import (
    SavedRouteCreated,
    SavedRouteCreateRequest,
    SavedRouteSnapshot,
)

router = APIRouter(prefix="/v2/saved-routes", tags=["saved routes"])


@router.post(
    "",
    response_model=SavedRouteCreated,
    status_code=status.HTTP_201_CREATED,
)
def create_saved_route(
    request: Annotated[SavedRouteCreateRequest, Body()],
    service: SavedRouteServiceDependency,
    response: Response,
) -> SavedRouteCreated:
    """Persist the exact request/candidate pair without invoking planning."""
    created = service.create(request)
    response.headers["Location"] = f"/v2/saved-routes/{created.slug}"
    response.headers["Cache-Control"] = "no-store"
    return created


@router.get("/{slug}", response_model=SavedRouteSnapshot)
def get_saved_route(
    slug: str,
    service: SavedRouteServiceDependency,
    response: Response,
) -> SavedRouteSnapshot:
    """Return one immutable unlisted snapshot without regeneration."""
    response.headers["Cache-Control"] = "private, no-store"
    response.headers["X-Robots-Tag"] = "noindex, nofollow, noarchive"
    response.headers["Referrer-Policy"] = "no-referrer"
    return service.get(slug)


@router.get("/{slug}/gpx", response_class=Response)
def get_saved_route_gpx(
    slug: str,
    service: SavedRouteServiceDependency,
) -> Response:
    """Serialize the stored candidate directly as a clean GPX track."""
    candidate = service.get(slug).candidate
    filename = gpx_filename(candidate.route.name)
    return Response(
        content=write_plan_gpx(candidate),
        media_type="application/gpx+xml",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
            "Cache-Control": "private, no-store",
            "X-Robots-Tag": "noindex, nofollow, noarchive",
            "Referrer-Policy": "no-referrer",
        },
    )


@router.delete("/{slug}", status_code=status.HTTP_204_NO_CONTENT)
def delete_saved_route(
    slug: str,
    service: SavedRouteServiceDependency,
    owner_token: Annotated[
        str | None,
        Header(alias="X-Saved-Route-Owner-Token"),
    ] = None,
) -> Response:
    """Delete a snapshot using the creation-time capability token."""
    service.delete(slug, owner_token)
    return Response(
        status_code=status.HTTP_204_NO_CONTENT,
        headers={"Cache-Control": "no-store"},
    )

File: src/sugarglider/web/static/saved_routes.js
import { ApiError } from "./api.js";

async function savedRouteError(response) {
  let body;
  try { body = await response.json(); } catch { body = null; }
  const error = body?.error;
  return new ApiError(
    error?.message ?? `The server returned HTTP ${response.status}.`,
    JSON.stringify({
      status: response.status,
      code: error?.code ?? "saved_route_error",
    }, null, 2),
    error?.code ?? "saved_route_error",
  );
}

export function sharedRouteSlug(pathname = window.location.pathname) {
  const match = pathname.match(/^\/r\/([A-Za-z0-9_-]{20,64})\/?$/);
  return match?.[1] ?? null;
}

export async function createSavedRoute(sourceRequest, candidate) {
  const response = await fetch("/v2/saved-routes", {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify({
      schema_version: 1,
      source_request: sourceRequest,
      candidate,
    }),
  });
  if (!response.ok) throw await savedRouteError(response);
  return response.json();
}

export async function getSavedRoute(slug) {
  const response = await fetch(`/v2/saved-routes/${encodeURIComponent(slug)}`, {
    headers: { Accept: "application/json" },
  });
  if (!response.ok) throw await savedRouteError(response);
  return response.json();
}

export async function deleteSavedRoute(slug, ownerToken) {
  if (!ownerToken) {
    throw new ApiError(
      "This browser does not hold the deletion capability for this saved route.",
      "",
      "saved_route_owner_token_missing",
    );
  }
  const response = await fetch(`/v2/saved-routes/${encodeURIComponent(slug)}`, {
    method: "DELETE",
    headers: { "X-Saved-Route-Owner-Token": ownerToken },
  });
  if (!response.ok) throw await savedRouteError(response);
}

export function savedRouteShareUrl(saved) {
  return new URL(
    saved.share_path ?? `/r/${saved.slug}`,
    window.location.origin,
  ).href;
}

export async function shareSavedRoute(saved) {
  if (typeof navigator.share !== "function") {
    throw new ApiError(
      "Sharing is not available in this browser. Copy the link instead.",
      "",
      "saved_route_share_unavailable",
    );
  }
  const url = savedRouteShareUrl(saved);
  await navigator.share({
    title: saved.route_name ?? "Sugarglider route",
    text: "View this immutable Sugarglider route snapshot.",
    url,
  });
}

export async function downloadSavedRouteGpx(slug) {
  const response = await fetch(
    `/v2/saved-routes/${encodeURIComponent(slug)}/gpx`,
    { headers: { Accept: "application/gpx+xml" } },
  );
  if (!response.ok) throw await savedRouteError(response);
  return {
    blob: await response.blob(),
    filename: attachmentFilename(
      response.headers.get("Content-Disposition"),
    ),
  };
}

function attachmentFilename(header) {
  const match = header?.match(/filename="?([^";]+)"?/i);
  return match?.[1] ?? "sugarglider-route.gpx";
}
